File: 03-trapping-rainwater/brute-force.py
def trapped_water(heights: list[int]) -> int:
    total_water = 0

    # Traverse the array from left to right
    for p in range(len(heights)):
        max_left = heights[p]
        max_right = heights[p]
        p_left = p
        p_right = p

        # Get the largest height on the left side of the current position
        while p_left >= 0:
            max_left = max(max_left, heights[p_left])
            p_left -= 1

        # Get the largest height on the right side of the current position
        while p_right < len(heights):
            max_right = max(max_right, heights[p_right])
            p_right += 1

        # Get the lowest height between the left and right side
        top = min(max_left, max_right)

        if top > heights[p]:
            total_water += top - heights[p]

    return total_water


heights = [0, 1, 0, 2, 1, 0, 3, 1, 0, 1, 2]

print(trapped_water(heights))

File: 03-trapping-rainwater/brute-force.ts
const trappedWater = (heights: number[]): number => {
    let totalWater = 0;

    // Traverse the array from left to right
    for (let p = 0; p < heights.length; p++) {
        let maxLeft: number = heights[p], maxRight: number = heights[p];
        let pLeft: number = p, pRight: number = p;

        // Get the largest height on the left side of the current position
        while (pLeft >= 0) {
            maxLeft = Math.max(maxLeft, heights[pLeft]);
            pLeft--;
        }

        // Get the largest height on the right side of the current position
        while (pRight < heights.length) {
            maxRight = Math.max(maxRight, heights[pRight]);
            pRight++;
        }

        // Get the lowest height between the left and right side
        const top = Math.min(maxLeft, maxRight);

        // Add water if current height is below the lowest height.
        if (top > heights[p]) {
            totalWater += top - heights[p];
        }
    }

    return totalWater;
}

const heights = [0, 1, 0, 2, 1, 0, 3, 1, 0, 1, 2];

console.log(trappedWater(heights));